Sprint 5/step 1

DemasX64 opened the pull request:
add feed page
add order page
add profile orders
add store types

DemasX64 committed:
add feed and order layouts

DemasX64 committed:
add profile orders layout

DemasX64 committed:
add feed page
add order page
add profile orders
add store types

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации

* при запуске проекта ошибка линтера https://disk.yandex.ru/i/BzTo3h66XXq-BQ
* Сейчас у Вас отображается **рандомный** номер заказа с сервера при отправке собранного бургера, а нужно порядковый номер заказа.  Теперь для получения **порядкового номера заказа** для отображения его в ленте и истории нужно отправлять заказ, добавляя в `headers` еще `Authorization` с Вашим токеном в значении. Спросите наставника об этом, если не получится.
* нет автологина. Без этого не получится проверить логику в **истории** заказов
* страница восстановления пароля не по заданию. Скрин https://disk.yandex.ru/i/gGVsxbcgJ-_U2g . Должен быть один инпут для ввода мэйла, после нажатии кнопки идет переход на `reset-password` для восстановления пароля (тоже один инпут)
*

DemasX64 committed:
fix forgot password page
fix autologin
fix create order req

gennady-bars commented on src/utils/auth-api.ts:
```diff
@@ -34,11 +34,11 @@ export const forgotPassword = createAsyncThunk<{email: string}, {}, {}>(
 
 export const resetPassword = createAsyncThunk<any, any, any>(
```
`createAsyncThunk<any, any, any>` не нужно указывать типы `any` везде в `createAsyncThunk`. Либо нужно удалить их, либо нужно типизировать правильно

gennady-bars commented on src/utils/api-service.ts:
```diff
@@ -1,12 +1,12 @@
+  // alert('Ошибка получения данных');
 };
 
 export const checkResponse = async (res: any, rejectWithValue: any) => {
```
тип `Response` для ответа от сервера можно использовать в `checkResponse `

```js

export const checkResponse = async (res: Response, rejectWithValue: (value: unknown) => void) => {


```

gennady-bars commented on src/pages/profile/profile.tsx:
```diff
@@ -69,7 +66,25 @@ const Profile = () => {
 
-  // const { name, email, password } = useSelector((state) => state.auth.user);
+  useEffect(() => {
+    const ws = new WebSocket(`wss://norma.nomoreparties.space/orders?token=${getCookie('accessToken')?.split(' ')[1]}`);
```
По заданию нужно сделать **socketMiddleware ** , а не просто в эффекте создавать сокет


В проекте должен быть только один **универсальный** `socketMiddleware`. Нужно передавать объект с типами в вызов для повышения универсальности и независимости от внешних данных. Передавать `url` можно через `payload`  с уже полными данными для отправки (токеном и тд). Посоветуйтесь с наставником

```js

export const socketMiddleware = (wsActions: TWsActions): Middleware => {

```

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -31,26 +32,32 @@ const App = () => {
-      }
-    })();
+  useEffect(() => {
+    const ws = new WebSocket('wss://norma.nomoreparties.space/orders/all');
```
* По заданию нужно сделать **socketMiddleware ** , а не просто в эффекте создавать сокет

gennady-bars replied:
Нельзя открывать соединения сокетов в `App` сразу при загрузке сайта. Пользователь даже может не зайти на страницы ленты или истории, а соединения будут открыты. Это лишняя нагрузка на сервер и приложение

gennady-bars replied:
При входе на страницу ленты заказов нужно **открывать** соединение сокета, а при покидании страницы – **закрывать**. Это делается все в одном хуке `useEffect`. То же самое и с историей заказов

gennady-bars commented on src/pages/profile/profile.tsx:
```diff
@@ -1,15 +1,17 @@
-import { NavLink } from 'react-router-dom';
-import { useStore } from 'react-redux';
+import { NavLink, Route } from 'react-router-dom';
+import { useSelector, useStore } from 'react-redux';
```
`import { useDispatch, useSelector } from 'react-redux';` такого больше не должно быть в проекте.  Хуки `useDispatch, useSelector` нужно типизировать **1 раз** в отдельном файле и заменить их во всем проекте на свои, а не из библиотеки `react-redux` .  Вот ссылка на документацию по их типизации https://redux.js.org/usage/usage-with-typescript#define-typed-hooks

gennady-bars commented on src/pages/profile/profile.tsx:
```diff
@@ -69,7 +66,25 @@ const Profile = () => {
+    };
+  }, []);
+
+  const orders = useSelector((state: RootState) => state.profileFeed.orders);
```
`(state: RootState)` уже не нужно будет типизировать вручную стейт, когда замените везде хук `useSelector`. Нужно будет удалить дублирование типизации

gennady-bars commented on src/pages/profile/profile.tsx:
```diff
@@ -90,41 +105,64 @@ const Profile = () => {
         <p className="text text_type_main-default text_color_inactive mt-20">В этом разделе вы можете изменить свои персональные данные</p>
       </div>
-      <div className={styles.fields}>
-        <Input
```
`if (e.keyCode === escKeyCode) {` в `Modal` уже **depricated**


Лучше использовать строки. `(evt.key === “Escape”)`.

gennady-bars commented on src/components/order/order.tsx:
```diff
@@ -0,0 +1,101 @@
+        </div>
+        <div className={footer}>
+          <div className={footerIngredients}>
+            {ingredientsObj.slice().reverse().map((ingredient, index:number) => {
```
Вам больше **не нужно типизировать** аргументы в `map`, `filter`, `reduce` и тд, если массив уже 1 раз типизирован. Тайпскрипт возьмет тип автоматически. Код станет чище

gennady-bars commented on src/components/order/order.tsx:
```diff
@@ -0,0 +1,101 @@
+        pathname: `/feed/${_id}`,
+        state: { background: location },
+      }}
+    >
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная, но есть некоторые недочеты: 


* в **истории** попап пустой открывается. Скрин https://disk.yandex.ru/i/seXiD9tITeD76w . Вы пытаетесь найти заказ в ленте, хотя кликаете на заказ в **истории**
* пользователь не должен **напрямую** попасть на маршрут `/reset-password`. На эту страницу **никакой** пользователь не должен попадать напрямую. Только после нажатия на `восстановить пароль` нужно делать переход на эту страницу. Для этого нужно сделать проверку, был ли пользователь на странице `forgot-password` перед тем, как попасть на `reset-password`
* `createAsyncThunk<any, any, any>` не нужно указывать типы `any` везде в `createAsyncThunk`. Либо нужно удалить их, либо нужно типизировать правильно
* тип `Response` для ответа от сервера можно использовать в `checkResponse `
* По заданию нужно сделать **socketMiddleware ** , а не просто в эффекте создавать сокет
* Нельзя открывать соединения сокетов в `App` сразу при загрузке сайта. Пользователь даже может не зайти на страницы ленты или истории, а соединения будут открыты. Это лишняя нагрузка на сервер и приложение
* При входе на страницу ленты заказов нужно **открывать** соединение сокета, а при покидании страницы – **закрывать**. Это делается все в одном хуке `useEffect`. То же самое и с историей заказов
* `import { useDispatch, useSelector } from 'react-redux';` такого больше не должно быть в проекте.  Хуки `useDispatch, useSelector` нужно типизировать **1 раз** в отдельном файле и заменить их во всем проекте на свои, а не из библиотеки `react-redux` .  Вот ссылка на документацию по их типизации https://redux.js.org/usage/usage-with-typescript#define-typed-hooks
* `(state: RootState)` уже не нужно будет типизировать вручную стейт, когда замените везде хук `useSelector`. Нужно будет удалить дублирование типизации
* в проекте не нужно использовать `useStore<RootState>`. Для получения данных из стора используется `useSelector` (типизированный)
* Вам больше **не нужно типизировать** аргументы в `map`, `filter`, `reduce` и тд, если массив уже 1 раз типизирован. Тайпскрипт возьмет тип автоматически. Код станет чище




**Можно лучше**

*  при клике на **логотип** обычно идет переход на главную страницу
*  желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

DemasX64 committed:
add socket middleware
fix order page in profile
hide reset password page
fix types of thunks
optimize socket connection
add typed useselector
some fixes

DemasX64 committed:
some fix

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -35,23 +34,6 @@ const App = () => {
-    };
-  }, []);
-
   const location = useLocation<any>();
```
Можно типизировать `location` вот так:

```jsx
export interface ILocation {
    from?: Location;
    background?: Location;
    pathname?: string;
}

// и передавать это в хук

const location = useLocation<ILocation>();


```

gennady-bars commented on src/middleware/socketMiddleware.ts:
```diff
@@ -0,0 +1,41 @@
+
+    return (next) => (action) => {
+      const { dispatch } = store;
+      if (action.type === 'orders/connect') {
```
`'orders/connect'`

это хардкод под определенную строку типа экшена, а нужно универсально все сделать, чтобы мидллвар подходил и для новостей с чатами. 



В проекте должен быть  **универсальный** `socketMiddleware`. Нужно передавать объект с типами в вызов для повышения универсальности и независимости от внешних данных. 

```js

export const socketMiddleware = (wsActions: TWsActions): Middleware => {

```

gennady-bars commented on src/middleware/socketMiddleware.ts:
```diff
@@ -0,0 +1,41 @@
+        socket = new WebSocket(action.payload);
+      }
+      if (socket) {
+        if (action.type === 'orders/disconnect') {
```
это хардкод под определенную строку типа экшена, а нужно универсально все сделать, чтобы мидллвар подходил и для новостей с чатами.

gennady-bars commented on src/middleware/socketMiddleware.ts:
```diff
@@ -0,0 +1,41 @@
+/* eslint-disable arrow-body-style */
+/* eslint-disable import/prefer-default-export */
+import type { Middleware, MiddlewareAPI } from 'redux';
+import { setData, setIsConnected } from '../services/reducers/orders';
```
`{ setData, setIsConnected } ` 


этот импорт нужно удалить

Нужно передавать объект с типами в вызов `socketMiddleware `

В проекте должен быть  **универсальный** `socketMiddleware`. Нужно передавать объект с типами в вызов для повышения универсальности и независимости от внешних данных. 

```js

export const socketMiddleware = (wsActions: TWsActions): Middleware => {

```

gennady-bars commented on src/pages/profile/profile.tsx:
```diff
@@ -67,23 +67,11 @@ const Profile = () => {
-    ws.onerror = () => {
-      dispatch(setIsConnected(false));
-    };
+    dispatch(connect(`wss://norma.nomoreparties.space/orders?token=${getCookie('accessToken')?.split(' ')[1]}`));
```
`wss://norma.nomoreparties.space/orders` дублируется несколько раз. Нужно вынести в константу `WS_BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

gennady-bars commented on src/components/order-page/order-page.tsx:
```diff
@@ -51,6 +55,15 @@ const OrderPage: FC<IOrderPage> = ({ setOrderNumber }) => {
 
+  useEffect(() => {
+    if (!setOrderNumber) {
+      dispatch(connect('wss://norma.nomoreparties.space/orders/all'));
```
`wss://norma.nomoreparties.space/orders` дублируется несколько раз. Нужно вынести в константу `WS_BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

gennady-bars commented on src/pages/feed-page/feed-page.tsx:
```diff
@@ -1,13 +1,21 @@
+  const dispatch = useAppDispatch();
+
+  useEffect(() => {
+    dispatch(connect('wss://norma.nomoreparties.space/orders/all'));
```
`wss://norma.nomoreparties.space/orders` дублируется несколько раз. Нужно вынести в константу `WS_BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.

gennady-bars reviewed (commented):
Уже почти все отлично, но:

* в 5ти местах не заменили хук `useDispatch` на свой https://disk.yandex.ru/i/iOh6w6XT-4CyZw 
* `'orders/connect'` это хардкод под определенную строку типа экшена, а нужно универсально все сделать, чтобы мидллвар подходил и для новостей с чатами. 
* В проекте должен быть  **универсальный** `socketMiddleware`. Нужно передавать объект с типами в вызов для повышения универсальности и независимости от внешних данных. 
* `wss://norma.nomoreparties.space/orders` дублируется несколько раз. Нужно вынести в константу `WS_BASE_URL` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было **сменить адрес**, но не менять эндпоинты.
* Кое-что не исправили https://github.com/DemasX64/react-burger/pull/7#discussion_r1025034437 
* В ленте почему-то дублируются данные в попапе и странице. СКрин https://disk.yandex.ru/i/kvie3965_xvchg . При первом открытии все ок, но через несколько секунд добавляется дублированнные ингредиенты. Иногда по 3 раза (друг под другом)
* В **истории** не открывается попап  заказа. Сразу страница показывается.  А нужна такая же логика, как в ленте

DemasX64 committed:
remove app dispatch

DemasX64 committed:
fix duplicate in order page

DemasX64 committed:
refactor deprecated keycode

DemasX64 committed:
add ws base url

DemasX64 committed:
remove hardcode from socketmiddleware

DemasX64 committed:
refactor socket middleware

DemasX64 committed:
refactor order page at profile orders

DemasX64 committed:
fix

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
